src/mlpack/core/optimizers/cmaes/cmaes.hpp

```diff
@@ -0,0 +1,398 @@
+/**
+ * @file cmaes.h
+ * @author Kartik Nighania (GSoC 17 mentor Marcus Edel)
+ *
+ * Covariance Matrix Adaptation Evolution Strategy
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_CMAES_CMAES_HPP
+#define MLPACK_CORE_OPTIMIZERS_CMAES_CMAES_HPP
+
+#include <cmath>
+#include <limits>
+#include <ostream>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <iostream>
+#include <armadillo>
+#include <iostream>
+
+#include "random.hpp"
+
+namespace mlpack {
+namespace optimization {
+//vector use of pop
+//cmaes as init
+/**
+ * @class CMAES 
+ * all the function parameters available to the user
+ */
+template<typename funcType, typename T>
+class CMAES
+{
+public:
+
+  //! constructor to initialize the algorithm parameters
+    CMAES(funcType& func, T *start = 0, T *stdDeviation = 0)
+  { 
+     size_t dimension = func.NumFunctions();
+     functionFitness = init(dimension, start, stdDeviation);
+  }
+
+  T Optimize(funcType& func, T *iterate);
```

Comment:
We should use ``arma::mat`` instead of a pointer here. That way we can basically reuse all mlpack functionality, without using ``memptr()`` or something similar. What do you think?